server/aws-lsp-codewhisperer/src/language-server/utilities/telemetryUtils.ts

```diff
@@ -0,0 +1,103 @@
+import { InitializeParams, ServerInfo } from '@aws/language-server-runtimes/server-interface'
+import { IdeCategory, UserContext } from '../../client/token/codewhispererbearertokenclient'
+
+const USER_AGENT_PREFIX = 'AWS-Language-Servers'
+
+export const getUserAgent = (initializeParams: InitializeParams, serverInfo?: ServerInfo): string => {
+    const format = (s: string) => s.replace(/\s/g, '-')
+
+    const items: String[] = []
+
+    items.push(USER_AGENT_PREFIX)
+
+    // Fields specific to runtime artifact
+    if (serverInfo?.name) {
+        serverInfo.version
+            ? items.push(`${format(serverInfo.name)}/${serverInfo.version}`)
+            : items.push(format(serverInfo.name))
+    }
+
+    // Compute client-specific suffix
+    // Missing required data fields are replaced with 'UNKNOWN' token
+    // Whitespaces in product.name and platform.name are replaced to '-'
+    if (initializeParams?.initializationOptions?.aws) {
+        const { clientInfo } = initializeParams?.initializationOptions?.aws
+        const { extension } = clientInfo || {}
+
+        if (extension) {
+            items.push(`${extension.name ? format(extension.name) : 'UNKNOWN'}/${extension.version || 'UNKNOWN'}`)
+        }
+
+        if (clientInfo) {
+            items.push(`${clientInfo.name ? format(clientInfo.name) : 'UNKNOWN'}/${clientInfo.version || 'UNKNOWN'}`)
+        }
+
+        if (clientInfo?.clientId) {
+            items.push(`ClientId/${clientInfo?.clientId}`)
+        }
+    } else {
+        // Default to standard InitializeParams.clientInfo if no custom aws.clientInfo is set
+        const { clientInfo } = initializeParams || {}
+        if (clientInfo) {
+            items.push(`${clientInfo.name ? format(clientInfo.name) : 'UNKNOWN'}/${clientInfo.version || 'UNKNOWN'}`)
+        }
+    }
+
+    return items.join(' ')
+}
+
+const IDE_CATEGORY_MAP: { [key: string]: IdeCategory } = {
```

Comment:
I'm looking for feedback on using custom extension name to get `IdeCategory`. 
Alternatives are:
1. Use `InitializeParams.clientInfo` set by IDE - since this value is outside of our control, we will need to maintain ever growing list of different variants to keep mapping up to date
2. Introduce new `IdeCategory` field in our custom `InitializeParams.initializationOptions.aws.clientInfo` and document enum of supported values.

Reply:
Both options not perfect:
1. We'll need to maintain the list, does not scale
2. IDEs need to inject CW-specific representation of IDE name, which should ideally be abstracted away
Do you think having both would make sense?

Reply:
Why are we mapping from `extension.name`, not `name`?

Reply:
`name` is meant to set name of IDE (or client destination), which is larger, more incoherent list, than `extension.name`. And typically it will be set from some hardcoded value set by IDE itself, which means that we (AWS) don't control this list.

`extension.name` is fully controlled by Toolkits and signifies toolkit extension name - smaller list, which we control.

Reply:
> I'd probably stick with 1, as I would not expect changes to happen often. And the values are anyway hardcoded on CW side.

Which option do you mean: the one I implemented to use `aws.clientInfo.extension.name` or `1. Use InitializeParams.clientInfo set by IDE`?

Reply:
Sorry, I meant the one you implemented. Ideally, CW should modify their model to make IDE category optional.

Reply:
> Ideally, CW should modify their model to make IDE category optional.

yeah, I discussed it internally with them, this might be done in future